### PR TITLE
Migrate Python C extension to Cython

### LINKED ISSUE #405: Replace all malloc/free calls with unique_ptr (or equivalent).
As we convert from C to C++, we should remove all calls to `malloc` and `free` and follow modern C++ best practices.

### PR DESCRIPTION
**Pull request by suquark**
## What do these changes do?

This PR migrates Python C extension to Cython bindings for the raylet client.

## Related issue number

#3474 
This also fixes #405

### CODE REVIEW

**Comment by suquark** [comment]
I have implemented all features in Cython. Some code still needs to be updated to fit the Cython interface.

File: python/ray/includes/unique_ids.pxi [hunk 1]
```diff
@@ -0,0 +1,903 @@
+"""This is a module for unique IDs in Ray.
+We define different types for different IDs for type safe.
+
+See https://github.com/ray-project/ray/issues/3721.
```
[the remark this answers is not in the document]
Comment (suquark) [reply python/ray/includes/unique_ids.pxi]: One reason I am defining those IDs as independent classes is because those C++ types could be strong-typedefed in the near future.
I have also tried to reuse some duplicated parts. Unfortunately, Cython cannot recognize classes from `cls` or `__class__` in many cases. So I still need to write a lot of extra code.

File: python/ray/function_manager.py [hunk 2]
```diff
@@ -377,8 +377,9 @@ def fetch_and_register_remote_function(self, key):
              "driver_id", "function_id", "name", "function", "num_return_vals",
              "module", "resources", "max_calls"
          ])
-        function_id = ray.ObjectID(function_id_str)
-        driver_id = ray.ObjectID(driver_id_str)
+        function_id = ray.FunctionID(function_id_str)
+        # TODO(suquark): Make sure the type of "driver_id".
```
Comment (pcmoritz) [inline python/ray/function_manager.py]: This is the right type, do you want to remove the TODO?

JobID and DriverID are the same, and in the future I'd say we should transition to using DriverID exclusively. But given that in C++ it's called JobID, I think it's ok to call it JobID in python too until we change it everywhere.

File: python/ray/includes/libraylet_client.pxd [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+from libcpp.utility cimport pair
+from libcpp.vector cimport vector as c_vector
+from libcpp.memory cimport unique_ptr
+from libcpp.unordered_map cimport unordered_map
+
+ctypedef unordered_map[c_string, c_vector[pair[int64_t, double]]] ResourceMappingType
+ctypedef pair[c_vector[ObjectID], c_vector[ObjectID]] WaitResultPair
+
+cdef extern from "ray/raylet/raylet_client.h" nogil:
+    cdef cppclass LibRayletClient "RayletClient":
```
Comment (pcmoritz) [inline python/ray/includes/libraylet_client.pxd]: just call this RayletClient or CRayletClient if it clashes?

File: python/ray/includes/task.pxd [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    cdef cppclass TaskExecutionSpecificationT:
+        c_vector[c_string] dependencies
+        double last_timestamp
+        int32_t num_forwards
+        TaskExecutionSpecificationT()
+
+    # Note: It is a C++ struct in the header file.
+    cdef cppclass TaskExecutionSpecification:
+        double last_timestamp() const
+        c_bool mutate_last_timestamp(double _last_timestamp)
```
Comment (pcmoritz) [inline python/ray/includes/task.pxd]: why the underscores here and below?

File: python/ray/includes/unique_ids.pxi (hunk 1 above)
[the remark this answers is not in the document]
Comment (pcmoritz) [reply python/ray/includes/unique_ids.pxi]: Do you see any other way to get rid of the code duplication? It would be good to resolve this before the PR is merged.

File: python/ray/libraylet_library_python.pyx [hunk 5]
```diff
@@ -0,0 +1,723 @@
+# cython: profile=False
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: how do you feel about calling this file just "raylet.pyx"? That seems to be what it is :)

File: python/ray/libraylet_library_python.pyx [hunk 6]
```diff
@@ -0,0 +1,723 @@
...
+        raise ValueError("The range of 'put_index' should be [1, %d]" % kMaxTaskPuts)
+    cdef CObjectID object_id = ComputePutId(task_id.data, put_index)
+    return PyObjectID.from_native(object_id)
+
+
+def compute_task_id(PyObjectID object_id):
+    return PyTaskID.from_native(ComputeTaskId(object_id.data))
+
+
+cdef c_bool is_simple_value(value, int *num_elements_contained):
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: why not cdef int num_elements_contained at the beginning and at the end return it (and get rid of the out argument)?

File: python/ray/libraylet_library_python.pyx [hunk 7]
```diff
@@ -0,0 +1,723 @@
...
+
+    def __repr__(self):
+        if <c_int32>self.lang == <c_int32>LANGUAGE_PYTHON:
+            return "PYTHON"
+        elif <c_int32>self.lang == <c_int32>LANGUAGE_CPP:
+            return "CPP"
+        elif <c_int32>self.lang == <c_int32>LANGUAGE_JAVA:
+            return "JAVA"
+        else:
+            raise Exception("Unexpected error")
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: something more descriptive like "language not recognized"?

File: python/ray/includes/unique_ids.pxi [hunk 8]
```diff
@@ -0,0 +1,986 @@
...
+            raise TypeError("Unsupported type: " + str(type(object_id)))
+
+    @staticmethod
+    cdef from_native(const CUniqueID& cpp_id):
+        cdef PyUniqueID self = PyUniqueID.__new__(PyUniqueID)
+        self.data = cpp_id
+        return self
+
+    @staticmethod
+    def from_random():
```
Comment (pcmoritz) [inline python/ray/includes/unique_ids.pxi]: Let's remove this method everywhere. As mentioned in the previous object id refactor PR, it's not fork-safe, so we have to be very careful about using it from python...

File: python/ray/libraylet_library_python.pyx [hunk 9]
```diff
@@ -0,0 +1,723 @@
...
+
+    def actor_id(self):
+        """Return the actor ID for this task."""
+        return PyActorID.from_native(self.task_spec.get().ActorId())
+
+    def actor_counter(self):
+        """Return the actor counter for this task."""
+        return self.task_spec.get().ActorCounter()
+
+    def _serialized_raylet_task(self):
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: How do you feel about leaving this in C++ and just wrapping it in Cython? Not sure if it makes a lot of sense to have all these Cython wrappers for flatbuffers.

File: python/ray/libraylet_library_python.pyx [hunk 10]
```diff
@@ -0,0 +1,723 @@
...
+            Offset[_Task] task_flatbuffer = task[0].ToFlatbuffer(fbb)
+        fbb.Finish(task_flatbuffer)
+        out_str = fbb.GetBufferPointer()[:fbb.GetSize()]
+        del task
+        del execution_spec
+        return out_str
+
+
+cdef class RayletClient:
+    cdef unique_ptr[LibRayletClient] client
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: yeah, as stated above, CRayletClient is the better name here I think

File: python/ray/libraylet_library_python.pyx [hunk 11]
```diff
@@ -0,0 +1,723 @@
...
+                  c_bool is_worker,
+                  PyJobID driver_id):
+        cdef:
+            c_string c_raylet_socket = to_c_string(raylet_socket)
+        # We have known that we are using Python, so just skip the language parameter.
+        self.client.reset(new LibRayletClient(c_raylet_socket, client_id.data,
+                          is_worker, driver_id.data, LANGUAGE_PYTHON))
+
+    def disconnect(self):
+        cdef:
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: having this on the same line without the colon makes sense if there is only one declaration (same above)

File: python/ray/libraylet_library_python.pyx [hunk 12]
```diff
@@ -0,0 +1,723 @@
...
+            CRayStatus status = self.client.get().Disconnect()
+        if not status.ok():
+            raise ConnectionError("[RayletClient] Failed to disconnect.")
+
+    def submit_task(self, Task task_spec):
+        cdef:
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: and here

File: python/ray/libraylet_library_python.pyx [hunk 13]
```diff
@@ -0,0 +1,723 @@
...
+            c_string c_raylet_socket = to_c_string(raylet_socket)
+        # We have known that we are using Python, so just skip the language parameter.
+        self.client.reset(new LibRayletClient(c_raylet_socket, client_id.data,
+                          is_worker, driver_id.data, LANGUAGE_PYTHON))
+
+    def disconnect(self):
+        cdef:
+            CRayStatus status = self.client.get().Disconnect()
+        if not status.ok():
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: we could have check_status similar to arrow here

File: python/ray/libraylet_library_python.pyx [hunk 14]
```diff
@@ -0,0 +1,723 @@
...
+            raise ConnectionError("[RayletClient] Failed to disconnect.")
+
+    def submit_task(self, Task task_spec):
+        cdef:
+            CRayStatus cstatus
+        cstatus = self.client.get().SubmitTask(task_spec.execution_dependencies.get()[0], task_spec.task_spec.get()[0])
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: gil can be cropped for this operation

File: python/ray/libraylet_library_python.pyx [hunk 15]
```diff
@@ -0,0 +1,723 @@
...
+            raise Exception(cstatus.ToString())
+
+    def wait(self, object_ids, int num_returns, c_int64 timeout_milliseconds,
+             c_bool wait_local, PyTaskID current_task_id):
+        cdef:
+            WaitResultPair result
+            c_vector[CObjectID] wait_ids
+            CRayStatus cstatus
+        wait_ids = ObjectIDsToVector(object_ids)
+        cstatus = self.client.get().Wait(wait_ids, num_returns, timeout_milliseconds,
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: drop gil

File: python/ray/libraylet_library_python.pyx [hunk 16]
```diff
@@ -0,0 +1,723 @@
...
+            c_vector[CObjectID] fetch_ids
+            CRayStatus cstatus
+        fetch_ids = ObjectIDsToVector(object_ids)
+        cstatus = self.client.get().FetchOrReconstruct(fetch_ids, fetch_only, current_task_id.data)
+        if not cstatus.ok():
+            raise Exception(cstatus.ToString())
+
+    def notify_unblocked(self, PyTaskID current_task_id):
+        cdef CRayStatus cstatus
+        cstatus = self.client.get().NotifyUnblocked(current_task_id.data)
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: drop gil

File: python/ray/libraylet_library_python.pyx [hunk 17]
```diff
@@ -0,0 +1,723 @@
...
+        if not cstatus.ok():
+            raise Exception(cstatus.ToString())
+
+    def fetch_or_reconstruct(self, object_ids,
+                             c_bool fetch_only, PyTaskID current_task_id=PyTaskID.nil()):
+        cdef:
+            c_vector[CObjectID] fetch_ids
+            CRayStatus cstatus
+        fetch_ids = ObjectIDsToVector(object_ids)
+        cstatus = self.client.get().FetchOrReconstruct(fetch_ids, fetch_only, current_task_id.data)
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: drop gil

File: python/ray/libraylet_library_python.pyx [hunk 18]
```diff
@@ -0,0 +1,723 @@
...
+            for i in range(c_value.size()):
+                ids_and_fractions.append((c_value[i].first, c_value[i].second))
+            resources_dict[key] = ids_and_fractions
+            postincrement(iterator)
+        return resources_dict
+
+    def push_error(self, PyJobID job_id, error_type, error_message,
+                   double timestamp):
+        cdef CRayStatus cstatus
+        cstatus = self.client.get().PushError(job_id.data,
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: drop gil

File: python/ray/libraylet_library_python.pyx (hunk 9 above)
Comment (pcmoritz) [reply python/ray/libraylet_library_python.pyx]: also then we can re-use it in different language implementations in the future

File: python/ray/libraylet_library_python.pyx [hunk 19]
```diff
@@ -0,0 +1,723 @@
...
+                    if profile_event.extra_data.length() == 0:
+                        raise ValueError("'extra_data' should not be a null string.")
+                else:
+                    raise ValueError("Unknown profile event key '%s'" % key_string)
+            # Note that profile_info.profile_events is a vector of unique pointers, so
+            # profile_event will be deallocated when profile_info goes out of scope.
+            # "emplace_back" of vector has not been supported by Cython
+            profile_info.profile_events.push_back(unique_ptr[GCSProfileEventT](profile_event))
+
+        cstatus = self.client.get().PushProfileEvents(profile_info)
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: gil

File: python/ray/libraylet_library_python.pyx [hunk 20]
```diff
@@ -0,0 +1,723 @@
...
+                                              to_c_string(error_type),
+                                              to_c_string(error_message),
+                                              timestamp)
+        if not cstatus.ok():
+            raise Exception(cstatus.ToString())
+
+    def push_profile_events(self, component_type, PyUniqueID component_id,
+                            node_ip_address, profile_data):
+        cdef:
+            GCSProfileTableDataT profile_info
```
Comment (pcmoritz) [inline python/ray/libraylet_library_python.pyx]: same as above, it would be better to keep this in C++ I think

**Comment by pcmoritz** [comment]
Great, looks like this is pretty far. I left a few comments that should be taken care of before we can merge this.

File: python/ray/function_manager.py (hunk 2 above)
Comment (robertnishihara) [reply python/ray/function_manager.py]: Yeah, we actually use both `DriverID` and `JobID` in C++. I agree about using only one (e.g., `DriverID`).

File: python/ray/includes/unique_ids.pxi (hunk 1 above)
[the remark this answers is not in the document]
Comment (robertnishihara) [reply python/ray/includes/unique_ids.pxi]: I agree that all of these types will be different C++ classes in the future, so we should plan for that setting. It's also possible that they will have different methods, e.g, with `ObjectID` you could potentially call `GetTaskID()` or `IsPut()`, but that wouldn't make sense with an `ActorID`.

File: python/ray/libraylet_library_python.pyx [hunk 21]
```diff
@@ -0,0 +1,723 @@
...
+        if value.dtype == "O":
+            return False
+        num_elements_contained[0] += value.nbytes
+        return num_elements_contained[0] < RayConfig.instance().num_elements_limit()
+
+    return False
+
+
+def check_simple_value(value):
+    """This method checks if a Python object is sufficiently simple that it can be
```
Comment (robertnishihara) [inline python/ray/libraylet_library_python.pyx]: Follow our usual Python docstring convention.

File: python/ray/includes/unique_ids.pxi (hunk 1 above)
[the remark this answers is not in the document]
Comment (pcmoritz) [reply python/ray/includes/unique_ids.pxi]: Agreed, but this doesn't mean they can't share code. I'm pretty sure with interitence it can be made to work well that the common methods are defined only once, I'll give it a shot.

File: python/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,49 @@
+# NOTE: These must be checked before including Cython packages to ensure we are using the right python version.
+# Segfaults could happen if we are using the wrong version.
+set(PYTHON_INCLUDE_DIR ${PYTHON_INCLUDE_DIRS})
+set(PYTHON_LIBRARY ${PYTHON_LIBRARIES})
+
+include(FindCython)
```
Comment (robertnishihara) [inline python/CMakeLists.txt]: `FindCython` runs `find_package( PythonInterp )`. However, in `setup.py`, when we call `build.sh`, we pass in the Python executable via

```python
subprocess.check_call(["../build.sh", "-p", sys.executable])
```

The Python we want to find is the one that is used to call `setup.py`. So, in the scenario where the Python used to call `setup.py` is different from the first Python on the user's PATH, will we find the right Cython executable?

Comment (pcmoritz) [reply python/CMakeLists.txt]: Good point, let me see if we can pass this through somehow

File: python/ray/_raylet.pyx [hunk 23]
```diff
@@ -0,0 +1,647 @@
...
+            actor_creation_dummy_object_id.data, max_actor_reconstructions, CUniqueID(actor_id.data),
+            CUniqueID(actor_handle_id.data), actor_counter, task_new_actor_handles, task_args, num_returns,
+            required_resources, required_placement_resources, LANGUAGE_PYTHON,
+            c_function_descriptor))
+
+        # Set the task's execution dependencies.
+        self.execution_dependencies.reset(new c_vector[CObjectID]())
+        if execution_arguments is not None:
+            for execution_arg in execution_arguments:
+                self.execution_dependencies.get().push_back((<ObjectID?>execution_arg).data)
```
Comment (robertnishihara) [inline python/ray/_raylet.pyx]: What does `<ObjectID?>` mean?

File: python/ray/_raylet.pyx [hunk 24]
```diff
@@ -0,0 +1,647 @@
...
+    def returns(self):
+        """Return the object IDs for the return values of the task."""
+        cdef RayletTaskSpecification *task_spec = self.task_spec.get()
+        return_id_list = []
+        for i in range(task_spec.NumReturns()):
+            return_id_list.append(ObjectID.from_native(task_spec.ReturnId(i)))
+        return return_id_list
+
+    def required_resources(self):
+        """Return the resource vector of the task."""
```
Comment (robertnishihara) [inline python/ray/_raylet.pyx]: It's actually a dictionary, not a vector.

File: python/ray/_raylet.pyx [hunk 25]
```diff
@@ -0,0 +1,647 @@
...
+        return self.task_spec.get().ActorCounter()
+
+
+cdef class RayletClient:
+    cdef unique_ptr[CRayletClient] client
+    def __cinit__(self, raylet_socket,
+                  ClientID client_id,
+                  c_bool is_worker,
+                  DriverID driver_id):
+        # We have known that we are using Python, so just skip the language parameter.
```
Comment (robertnishihara) [inline python/ray/_raylet.pyx]: `We have known` -> `We know`

File: python/ray/_raylet.pyx [hunk 26]
```diff
@@ -0,0 +1,647 @@
...
+        self.client.reset(new CRayletClient(raylet_socket.encode("ascii"), client_id.data,
+                          is_worker, driver_id.data, LANGUAGE_PYTHON))
+
+    def disconnect(self):
+        check_status(self.client.get().Disconnect())
+
+    def submit_task(self, Task task_spec):
+        check_status(self.client.get().SubmitTask(task_spec.execution_dependencies.get()[0], task_spec.task_spec.get()[0]))
```
Comment (robertnishihara) [inline python/ray/_raylet.pyx]: Instead of using `[0]`, shouldn't we dereference the pointers? Since `[0]` gives the impression that it is an array.

Comment (robertnishihara) [inline python/ray/_raylet.pyx]: Instead of using `[0]`, shouldn't we dereference the pointers? Since `[0]` gives the impression that it is an array.

File: python/ray/_raylet.pyx [hunk 27]
```diff
@@ -0,0 +1,647 @@
...
+                                                 timestamp))
+
+    def push_profile_events(self, component_type, UniqueID component_id,
+                            node_ip_address, profile_data):
+        cdef:
+            GCSProfileTableDataT profile_info
+            GCSProfileEventT *profile_event
+            c_string event_type
+
+        if not profile_data:
```
Comment (robertnishihara) [inline python/ray/_raylet.pyx]: Let's use `if len(profile_data) == 0` instead

File: python/ray/includes/unique_ids.pxd [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+        uint8_t *mutable_data();
+        size_t size() const;
+        c_string binary() const;
+        c_string hex() const;
+
+    ctypedef CUniqueID TaskID
+    ctypedef CUniqueID ObjectID
+    ctypedef CUniqueID JobID
+    ctypedef CUniqueID FunctionID
+    ctypedef CUniqueID ClassID
```
Comment (robertnishihara) [inline python/ray/includes/unique_ids.pxd]: I think `ActorClassID` would be a better name.

File: python/ray/includes/common.pxd [hunk 29]
```diff
@@ -0,0 +1,155 @@
...
+        void AddResources(const ResourceSet &other)
+        c_bool SubtractResourcesStrict(const ResourceSet &other)
+        c_bool GetResource(const c_string &resource_name, double *value) const
+        double GetNumCpus() const
+        c_bool IsEmpty() const
+        const unordered_map[c_string, double] &GetResourceMap() const
+        const c_string ToString() const
+
+
+cdef inline object PyObject_to_object(PyObject*o):
```
Comment (robertnishihara) [inline python/ray/includes/common.pxd]: `PyObject *o`

File: python/ray/__init__.py [hunk 30]
```diff
@@ -49,7 +49,9 @@
 modin_path = os.path.join(os.path.abspath(os.path.dirname(__file__)), "modin")
 sys.path.append(modin_path)
 
-from ray.raylet import ObjectID, _config  # noqa: E402
+from ray.raylet import (UniqueID, ObjectID, DriverID, ClientID,  # noqa: E402
+                        ActorID, ActorHandleID, FunctionID,  # noqa: E402
+                        ClassID, TaskID, _config)  # noqa: E402
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply python/ray/__init__.py]: fixed

File: cmake/Modules/UseCython.cmake [hunk 31]
```diff
@@ -0,0 +1,320 @@
+# Define a function to create Cython modules.
```
[the remark this answers is not in the document]
Comment (suquark) [reply cmake/Modules/UseCython.cmake]: They are from https://github.com/thewtex/cython-cmake-example.

File: python/ray/_raylet.pyx (hunk 26 above)
Comment (suquark) [reply python/ray/_raylet.pyx]: They are common in the official doc as an alternative to "*": https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#types

File: python/ray/_raylet.pyx (hunk 23 above)
Comment (suquark) [reply python/ray/_raylet.pyx]: Cython will check and convert the type of the object. If it is not a ObjectID, Cython will raise a TypeError; otherwise Cython will treat it as a Cython `ObjectID` object (it is different from a Python object because we can access its native components and it is statically compiled).

File: python/ray/_raylet.pyx [hunk 32]
```diff
@@ -0,0 +1,647 @@
...
+
+def compute_task_id(ObjectID object_id):
+    return TaskID.from_native(ComputeTaskId(object_id.data))
+
+
+cdef c_bool is_simple_value(value, int *num_elements_contained):
+    # Because "RayConfig" have a private destructor,
+    # we need to call "RayConfig.instance()" instead of defining a new variable.
+
+    num_elements_contained[0] += 1
```
[the remark this answers is not in the document]
Comment (suquark) [reply python/ray/_raylet.pyx]: Because `*` means "unpack" in Cython. The official doc suggests that we can use `[0]` as an alternative in most cases.

File: python/ray/includes/task.pxd (hunk 4 above)
Comment (suquark) [reply python/ray/includes/task.pxd]: Because it underscores in the C++ code.

File: python/ray/_raylet.pyx (hunk 32 above)
[the remark this answers is not in the document]
Comment (robertnishihara) [reply python/ray/_raylet.pyx]: Ok, I see. Thanks.

File: python/ray/libraylet_library_python.pyx (hunk 21 above)
Comment (pcmoritz) [reply python/ray/libraylet_library_python.pyx]: done

File: python/CMakeLists.txt (hunk 22 above)
Comment (pcmoritz) [reply python/CMakeLists.txt]: done

File: python/ray/_raylet.pyx [hunk 33]
```diff
@@ -0,0 +1,647 @@
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+from ray.includes.common cimport *
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply python/ray/_raylet.pyx]: removed it everywhere

File: python/ray/_raylet.pyx [hunk 34]
```diff
@@ -0,0 +1,647 @@
...
+    if status.ok():
+        return 0
+
+    with gil:
+        message = str(status.message())
+        raise Exception(message)
+
+
+cdef c_vector[CObjectID] ObjectIDsToVector(object_ids):
+    """A helper function that converts a Python list of object ids to a vector.
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply python/ray/_raylet.pyx]: yeah, let's do it in a subsequent PR

[the remark this answers is not in the document]
Comment (pcmoritz) [reply python/ray/_raylet.pyx]: done

File: python/ray/_raylet.pyx [hunk 35]
```diff
@@ -0,0 +1,647 @@
...
+        if value.dtype == "O":
+            return False
+        num_elements_contained[0] += value.nbytes
+        return num_elements_contained[0] < RayConfig.instance().num_elements_limit()
+
+    return False
+
+
+def check_simple_value(value):
+    """This method checks if a Python object is sufficiently simple that it can be
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply python/ray/_raylet.pyx]: done

File: python/ray/_raylet.pyx [hunk 36]
```diff
@@ -0,0 +1,323 @@
...
+        raise ValueError("The range of 'put_index' should be [1, %d]" % kMaxTaskPuts)
+    return ObjectID.from_native(ComputePutId(task_id.data, put_index))
+
+
+def compute_task_id(ObjectID object_id):
+    return TaskID.from_native(ComputeTaskId(object_id.data))
+
+
+cdef c_bool is_simple_value(value, int *num_elements_contained):
+    # Because "RayConfig" have a private destructor,
```
Comment (robertnishihara) [inline python/ray/_raylet.pyx]:
```suggestion
    # Because "RayConfig" has private destructor,
```

Comment (robertnishihara) [reply python/ray/_raylet.pyx]: though i'm not sure how necessary this comment is. This is just the intended usage of `RayConfig` since it's a singleton.

File: python/ray/_raylet.pyx [hunk 37]
```diff
@@ -0,0 +1,323 @@
...
+
+    if cpython.PyBytes_CheckExact(value):
+        num_elements_contained[0] += cpython.PyBytes_Size(value)
+        return num_elements_contained[0] < RayConfig.instance().num_elements_limit()
+
+    if cpython.PyUnicode_CheckExact(value):
+        num_elements_contained[0] += cpython.PyUnicode_GET_SIZE(value)
+        return num_elements_contained[0] < RayConfig.instance().num_elements_limit()
+
+    if cpython.PyList_CheckExact(value) and cpython.PyList_Size(value) < RayConfig.instance().size_limit():
```
Comment (robertnishihara) [inline python/ray/_raylet.pyx]: It should be a separate PR, but we should follow this up with linting for the Cython files soon.

File: python/ray/_raylet.pyx [hunk 38]
```diff
@@ -0,0 +1,323 @@
...
+        for item in value:
+            if not is_simple_value(item, num_elements_contained):
+                return False
+        return num_elements_contained[0] < RayConfig.instance().num_elements_limit()
+
+    if cpython.PyDict_CheckExact(value) and cpython.PyDict_Size(value) < RayConfig.instance().size_limit():
+        # TODO(suquark): Use "items" in Python2 would be not very efficient.
```
Comment (robertnishihara) [inline python/ray/_raylet.pyx]:
```suggestion
        # TODO(suquark): Using "items" in Python2 is not very efficient.
```

File: python/ray/includes/ray_config.pxd [hunk 39]
```diff
@@ -0,0 +1,83 @@
+from libc.stdint cimport int64_t, uint64_t
+from libcpp.string cimport string as c_string
+from libcpp.unordered_map cimport unordered_map
+
+
+cdef extern from "ray/ray_config.h" nogil:
```
Comment (robertnishihara) [inline python/ray/includes/ray_config.pxd]: Is the plan to simply keep this in sync with `ray_config.h`? If so, can we put a prominent note near the top of `ray_config.h` saying 

```
IF YOU MODIFY THIS FILE and add a configuration parameter, you must change
at least three additional things:
    1. You must update the "initialize" method in the "RayConfig" class.
    2. You must update the file "ray/python/ray/includes/ray_config.pxd".
    3. You must update the file "ray/python/ray/includes/ray_config.pxi".
```

File: python/ray/includes/ray_config.pxi [hunk 40]
```diff
@@ -0,0 +1,146 @@
...
+    def object_manager_default_chunk_size():
+        return RayConfig.instance().object_manager_default_chunk_size()
+
+    @staticmethod
+    def num_workers_per_process():
+        return RayConfig.instance().num_workers_per_process()
+
+    @staticmethod
+    def max_task_lease_timeout_ms():
+        return RayConfig.instance().max_task_lease_timeout_ms()
```
Comment (robertnishihara) [inline python/ray/includes/ray_config.pxi]: new line at end

File: python/ray/includes/task.pxi [hunk 41]
```diff
@@ -0,0 +1,208 @@
...
+        cdef:
+            unordered_map[c_string, double] resource_map = self.task_spec.get().GetRequiredResources().GetResourceMap()
+            c_string resource_name
+            double resource_value
+            unordered_map[c_string, double].iterator iterator = resource_map.begin()
+
+        required_resources = {}
+        while iterator != resource_map.end():
+            resource_name = dereference(iterator).first
+            # TODO(suquark): What is the type of the resource name (bytes, str, unicode)?
```
Comment (robertnishihara) [inline python/ray/includes/task.pxi]: In Python 2 it is required to be str (bytes). In Python 3 it is required to by str (unicode).

File: python/ray/includes/unique_ids.pxi [hunk 42]
```diff
@@ -0,0 +1,280 @@
+"""This is a module for unique IDs in Ray.
+We define different types for different IDs for type safe.
```
Comment (robertnishihara) [inline python/ray/includes/unique_ids.pxi]:
```suggestion
We define different types for different IDs for type safety.
```

File: src/ray/gcs/tables.h [hunk 43]
```diff
@@ -513,7 +513,8 @@ class ProfileTable : private Log<UniqueID, ProfileTableData> {
   std::string DebugString() const;
 };
 
-using CustomSerializerTable = Table<ClassID, CustomSerializerData>;
+// TODO(suquark): Is the change from "ClassID" to "ActorClassID" reasonable here?
```
Comment (robertnishihara) [inline src/ray/gcs/tables.h]: Actually no it is not, `CustomSerializerTable` is currently unused I believe, but it is actually intended to keep track of custom serializers for arbitrary Python classes, not just actor classes. Since we got rid of `ClassID`, we can probably just use `UniqueID` here.

File: src/ray/raylet/task.cc [hunk 44]
```diff
@@ -46,6 +46,16 @@ void Task::CopyTaskExecutionSpec(const Task &task) {
   ComputeDependencies();
 }
 
+std::string TaskToFlatbuffer(const std::vector<ObjectID> *dependencies,
+                             const TaskSpecification *task_spec) {
```
Comment (robertnishihara) [inline src/ray/raylet/task.cc]: Why not make this a method of `Task`?

Comment (robertnishihara) [reply src/ray/raylet/task.cc]: Also, maybe `SerializeAsString` or `SerializeTaskAsString`? It might seem like this method converts the task to a flatbuffer object.

File: src/ray/raylet/task_spec.h [hunk 45]
```diff
@@ -155,6 +155,13 @@ class TaskSpecification {
   flatbuffers::Offset<flatbuffers::String> ToFlatbuffer(
       flatbuffers::FlatBufferBuilder &fbb) const;
 
+  std::string ToFlatbuffer() const {
```
Comment (robertnishihara) [inline src/ray/raylet/task_spec.h]: `SerializeAsFlatbufferString`? Or `SerializeAsString` or something like that?

File: test/actor_test.py [hunk 46]
```diff
@@ -331,7 +331,7 @@ def __init__(self):
     random.seed(1234)
     f2 = Foo.remote()
 
-    assert f1._ray_actor_id.id() != f2._ray_actor_id.id()
+    assert f1._ray_actor_id.binary() != f2._ray_actor_id.binary()
```
Comment (robertnishihara) [inline test/actor_test.py]: We can just compare the IDs directly, right?

File: python/ray/_raylet.pyx (hunk 36 above)
Comment (pcmoritz) [reply python/ray/_raylet.pyx]: I removed it

File: python/ray/_raylet.pyx (hunk 37 above)
Comment (pcmoritz) [reply python/ray/_raylet.pyx]: I added an issue: https://github.com/ray-project/ray/issues/3829

File: python/ray/includes/ray_config.pxd (hunk 39 above)
Comment (pcmoritz) [reply python/ray/includes/ray_config.pxd]: done

File: python/ray/includes/ray_config.pxi (hunk 40 above)
Comment (pcmoritz) [reply python/ray/includes/ray_config.pxi]: done

File: python/ray/_raylet.pyx (hunk 38 above)
Comment (pcmoritz) [reply python/ray/_raylet.pyx]: done

File: python/ray/includes/unique_ids.pxi (hunk 42 above)
Comment (pcmoritz) [reply python/ray/includes/unique_ids.pxi]: done

File: src/ray/gcs/tables.h (hunk 43 above)
Comment (pcmoritz) [reply src/ray/gcs/tables.h]: using UniqueID now

File: src/ray/raylet/task.cc (hunk 44 above)
Comment (pcmoritz) [reply src/ray/raylet/task.cc]: I renamed it to SerializeTaskAsString; we cannot make this a method of Task because we cannot instantiate a Task in Cython on the stack at the moment (because it has a non-trivial constructor). We could make it static, but I think the current solution is good too.

File: src/ray/raylet/task_spec.h (hunk 45 above)
Comment (pcmoritz) [reply src/ray/raylet/task_spec.h]: I'm doing SerializeAsString to be consistent with SerializeTaskAsString

File: test/actor_test.py (hunk 46 above)
Comment (pcmoritz) [reply test/actor_test.py]: yeah, changed it

**Comment by suquark** [comment]
jenkins, retest please

**Comment by suquark** [comment]
Ready to merge. The CI errors come from the master branch.

**Review by pcmoritz — APPROVED** [review]
